Client/Frontend/Home/TopSitesPanel.swift

```diff
@@ -119,36 +119,38 @@ class TopSitesPanel: UIViewController {
             self.dataSource.data = data
             self.dataSource.profile = self.profile
 
-            // redraw now we've udpated our sources
+            // redraw now we've updated our sources
             self.collection?.collectionViewLayout.invalidateLayout()
             self.collection?.setNeedsLayout()
         }
     }
 
-    private func updateRemoveButtonStates() {
-        for i in 0..<layout.thumbnailCount {
-            if let cell = collection?.cellForItemAtIndexPath(NSIndexPath(forItem: i, inSection: 0)) as? ThumbnailCell {
-                //TODO: Only toggle the remove button for non-suggested tiles for now
-                if i < dataSource.data.count {
-                    cell.toggleRemoveButton(editingThumbnails)
-                } else {
-                    cell.toggleRemoveButton(false)
-                }
-            }
-        }
+    private func updateAllRemoveButtonStates() {
+        collection?.indexPathsForVisibleItems().forEach { updateRemoveButtonStateForIndexPath($0) }
     }
 
     private func deleteHistoryTileForSite(site: Site, atIndexPath indexPath: NSIndexPath) {
-        func reloadThumbnails() {
-            self.profile.history.getTopSitesWithLimit(self.layout.thumbnailCount)
-                .uponQueue(dispatch_get_main_queue()) { result in
-                    self.deleteOrUpdateSites(result, indexPath: indexPath)
+        let mainQueue = dispatch_get_main_queue()
+        collection?.userInteractionEnabled = false
+        profile.history.removeSiteFromTopSites(site)
+            .bindQueue(mainQueue) { _ in
+                return self.profile.history.refreshTopSitesCache()
+            }.bindQueue(mainQueue) { _ in
+                return self.profile.history.getTopSitesWithLimit(self.maxFrecencyLimit)
```

Comment:
And this doesn't need to be on the main queue, so I think this whole method can be:

```
let newSites = self.profile.history
                   .removeSiteFromTopSites(site)    // Invalidates cache.
           >>> { self.profile.history.getTopSitesWithLimit(self.maxFrecencyLimit) }

return newSites.bindQueue(mainQueue) { result in
    self.deleteOrUpdateSites(result, indexPath: indexPath)
    self.collection?.userInteractionEnabled = true
}
```